mux.js

```diff
@@ -0,0 +1,248 @@
+var protocol = require('hypercore-protocol')
+var readify = require('./ready')
+var inherits = require('inherits')
+var events = require('events')
+var debug = require('debug')('multifeed')
+var hypercore = require('hypercore')
+var xtend = require('xtend')
+
+// constants
+var MULTIFEED = 'MULTIFEED'
+var PROTOCOL_VERSION = '2.0.0'
+// extensions
+var MANIFEST = 'MANIFEST'
+var REQUEST_FEEDS = 'REQUEST_FEEDS'
+/*
+var ANNOUNCE_FEED = 'ANNOUNCE_FEED'
+var REQUEST_FEED_SIGNATURE = 'REQUEST_FEED_SIGNATURE'
+var FEED_SIGNATURE = 'FEED_SIGNATURE'
+*/
+
+var SupportedExtensions = [
+  MANIFEST,
+  REQUEST_FEEDS
+  //ANNOUNCE_FEED
+  //REQUEST_MANIFEST,
+  //REQUEST_FEED_SIGNATURE,
+  //FEED_SIGNATURE,
+]
+
+// `key` - protocol encryption key
+// `opts`- hypercore-protocol opts
+function Multiplexer (key, opts) {
+  if (!(this instanceof Multiplexer)) return new Multiplexer(key, opts)
+  debug('[REPLICATION] New mux initialized', key.toString('hex'), opts)
+  var self = this
+  self._opts = opts = opts || {}
+  self.extensions = opts.extensions = SupportedExtensions || opts.extensions
+
+  // initialize
+  self._localHave = null
+  self._localWant = null
+  self._remoteHas = null
+  self._remoteWants = null
+
+  var stream = this.stream = protocol(Object.assign(opts,{
+    userData: Buffer.from(JSON.stringify({
+      client: MULTIFEED,
+      version: PROTOCOL_VERSION,
+      // Help! exposing available extensions might be good for future version tolerance,
+      // but at the same time we're poentiallyleaking our user-agent fingerprint to a third party.
+      extensions: self.extensions
+    }))
+  }))
+
+  var feed = this._feed = stream.feed(key)
+
+  stream.on('handshake', function () {
+    var header = JSON.parse(this.userData.toString('utf8'))
+    debug('[REPLICATION] recv\'d header: ', JSON.stringify(header))
+    if (!compatibleVersions(header.version, PROTOCOL_VERSION)) {
+      debug('[REPLICATION] aborting; version mismatch (us='+PROTOCOL_VERSION+')')
+      self._finalize(new Error('protocol version mismatch! us='+PROTOCOL_VERSION + ' them=' + header.version))
+      return
+    }
+
+    if (header.client != MULTIFEED) {
+      debug('[REPLICATION] aborting; Client mismatch! expected ', MULTIFEED, 'but got', header.client)
+      self._finalize(new Error('Client mismatch! expected ' + MULTIFEED + ' but got ' + header.client))
+      return
+    }
+    self.remoteClient = header
```

Comment:
I don't think this is used anywhere else. Can we drop it?

Reply:
Dropped it.